### PR TITLE
Fix FLIRT v5 file format decompression

### LINKED ISSUE #417: IDA .sig files with version 5 compression not supported
### Work environment

| Questions                                            | Answers
|------------------------------------------------------|--------------------
| OS/arch/bits (mandatory)                             | -
| File format of the file you reverse (mandatory)      | -
| Architecture/bits of the file (mandatory)            | -
| `rizin -v` full output, **not truncated** (mandatory)  | https://github.com/rizinorg/rizin/commit/012d3e96c2d586966e4d63e09102f3936506d27d

### Expected behavior

Open and dump the FLIRT file as usual.

### Actual behavior

Go to the `sig/` directory of the IDA Pro installation (any 7.x version would suffice):
```
$ rizin -
[0x00000000]> zfd ./pc/bcb5rt.sig
Sorry we do not support the signatures version 5 compression.
We encountered an error while parsing the file ./pc/bcb5rt.sig. Sorry.
```

It's implemented in:
- `librz/analysis/flirt.c` (function `flirt_parse()`)
- `librz/ore/cmd_zign.c`
- See also `doc/flirt.md`

### PR DESCRIPTION
**Pull request by b1llow**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

...

**Test plan**

Go to the sig/ directory of the IDA Pro installation (any 7.x version would suffice):

```
$ rizin -
[0x00000000]> zfd ./pc/bcb5rt.sig
01:
 0000:
  00:
   000008006368617200:
    0. 00 0000 000D 0000:@$xt$c
   ................00000000:
    0. 00 0000 0010 0000:@Forms@_16890
  03003000FFFFFFFF030000004C00500000000000000000000000000002:
   0. 07 06B5 005F 0000:?
 03:
  4844430500000000FFFFFFFF:
   0. 00 0000 000E 0000:?
  696E740400000080FFFFFF7F:
   0. 00 0000 000E 0000:?
...
```

**Closing issues**

closes #417

### CODE REVIEW

File: librz/util/zip.c [hunk 1]
```diff
@@ -24,6 +24,13 @@ static const char *gzerr(int n) {
 }
 
 RZ_API ut8 *rz_inflate(const ut8 *src, int srcLen, int *srcConsumed, int *dstLen) {
```
Comment (wargio) [inline librz/util/zip.c:30]: Rename this as `rz_inflate_ignore_header` and add a comment in the C file saying that this is to ignore if is zlib compressed or gzipped.

File: librz/util/zip.c [hunk 2]
```diff
@@ -24,6 +24,13 @@ static const char *gzerr(int n) {
...
+	return rz_inflatew(src, srcLen, srcConsumed, dstLen, MAX_WBITS + 32);
+}
+RZ_API ut8 *rz_inflate_(const ut8 *src, int srcLen, int *srcConsumed, int *dstLen) {
```
Comment (wargio) [inline librz/util/zip.c]: Rename this as `rz_inflate`

**Review by wargio — CHANGES_REQUESTED** [review]

File: librz/util/zip.c (hunk 2 above)
Comment (b1llow) [reply librz/util/zip.c]: ok

File: librz/include/rz_util/rz_file.h [hunk 3]
```diff
@@ -43,11 +43,13 @@ RZ_API char *rz_file_relpath(const char *base, const char *path);
...
 RZ_API char *rz_file_path_unix_to_local(const char *path);
 RZ_API char *rz_file_binsh(void);
+RZ_API ut8 *rz_inflatew(const ut8 *src, int srcLen, int *srcConsumed, int *dstLen, int wbits);
 RZ_API ut8 *rz_inflate(const ut8 *src, int srcLen, int *srcConsumed, int *dstLen);
+RZ_API ut8 *rz_inflate_ignore_header(const ut8 *src, int srcLen, int *srcConsumed, int *dstLen);
 RZ_API ut8 *rz_file_gzslurp(const char *str, int *outlen, int origonfail);
 RZ_API char *rz_stdin_slurp(int *sz);
 RZ_API RZ_OWN char *rz_file_slurp(const char *str, RZ_NULLABLE size_t *usz);
-//RZ_API char *rz_file_slurp_range(const char *str, ut64 off, ut64 sz);
+// RZ_API char *rz_file_slurp_range(const char *str, ut64 off, ut64 sz);
```
Comment (XVilka) [inline librz/include/rz_util/rz_file.h]: Probably better to remove this line

File: librz/analysis/flirt.c [hunk 4]
```diff
@@ -1389,12 +1390,15 @@ static RzFlirtNode *flirt_parse(const RzAnalysis *analysis, RzBuffer *flirt_buf)
 
 	if (header->features & IDASIG__FEATURE__COMPRESSED) {
 		if (version == 5) {
```
Comment (XVilka) [inline librz/analysis/flirt.c]: I guess `>= 5`

**Review by XVilka — CHANGES_REQUESTED** [review]
I know there are exist signatures of version 6 and higher. Do they work with the same code？

File: librz/analysis/flirt.c (hunk 4 above)
Comment (b1llow) [reply librz/analysis/flirt.c]: No, just version 5 is compressed without header.

Comment (wargio) [reply librz/analysis/flirt.c]: if there is a version 6 or higher, then is better to have a check on this

File: librz/util/zip.c [hunk 5]
```diff
@@ -23,7 +23,32 @@ static const char *gzerr(int n) {
 	return errors[n];
 }
 
+/**
+ * @brief inflare zlib compressed or gzipped, automatically accepts either the zlib or gzip format, and use MAX_WBITS as the window size logarithm.
```
Comment (XVilka) [inline librz/util/zip.c]: A minor typo: `inflare` -> `inflate`.

File: librz/util/zip.c [hunk 6]
```diff
@@ -23,7 +23,32 @@ static const char *gzerr(int n) {
...
+ * @see rz_inflatew()
+ */
 RZ_API ut8 *rz_inflate(const ut8 *src, int srcLen, int *srcConsumed, int *dstLen) {
+	return rz_inflatew(src, srcLen, srcConsumed, dstLen, MAX_WBITS + 32);
+}
+
+/**
+ * @brief inflare zlib compressed or gzipped. The input must be a raw stream with no header or trailer.
```
Comment (XVilka) [inline librz/util/zip.c]: Same here

File: librz/util/zip.c [hunk 7]
```diff
@@ -23,7 +23,32 @@ static const char *gzerr(int n) {
...
+ * @see rz_inflatew()
+ */
+RZ_API ut8 *rz_inflate_ignore_header(const ut8 *src, int srcLen, int *srcConsumed, int *dstLen) {
+	return rz_inflatew(src, srcLen, srcConsumed, dstLen, -MAX_WBITS);
+}
+
+/**
+ * \brief inflare zlib compressed or gzipped.
```
Comment (XVilka) [inline librz/util/zip.c]: And here

**Review by XVilka — CHANGES_REQUESTED** [review]
Minor things, it fixed some of the FLIRT singature files but not everything.
See the files `ida/sig/arm/armlibl.sig` and `ida/sig/arm/armlibb.sig`:
```
[0x00000000]> zfd ./armlibl.sig
inflate error: -3 data error
Decompressing failed.
We encountered an error while parsing the file. Sorry.
[0x00000000]> zfd armlibb.sig
inflate error: -3 data error
Decompressing failed.
We encountered an error while parsing the file. Sorry.
[0x00000000]> 
```

With file `ida/sig/mips/psx.sig` there is UB:
```
ida-7.6-sp1/sig/mips took 8s 
[i] ℤ rizin =                                                                                                                                                                                                                     
[0x00000000]> zfd psx.sig
../librz/analysis/flirt.c:383:26: runtime error: left shift of 52428 by 16 places cannot be represented in type 'int'
....023C....4224....033C....6324000040AC040042242B084300FCFF2014:
 0. 04 DEFC 00A8 0000:o=20:a=12A:psyq:m=+A0^/main (0087: 00)
 1. 04 DEFC 0180 0000:o=20:a=12A:psyq:m=+9C^/main 00A8:__main 0118:__do_global_dtors (0087: 3C)
[0x00000000]> 
```

I tried these files from different versions of IDA Pro - 7.2, 7.5, 7.6 - the result is the same, they seem not changed much.

**Comment by b1llow** [comment]
This is version 6, it seems to be a old bug. Maybe need a little more time.

```
[0x00000000]> zfd ../sig_75/arm/armlibl.sig
version: 6
arch: 386
file_types: DOS_EXE_OLD DOS_COM_OLD BIN DOSDRV NE INTELHEX MOSHEX LX LE NLM COFF PE OMF SREC ZIP OMFLIB AR LOADER ELF W32RUN AOUT PILOT EXE AIXAR
os_types: MSDOS WIN OS2 NETWARE UNIX
app_types: CONSOLE GRAPHICS EXE DLL DRV SINGLE_THREADED MULTI_THREADED 16_BIT 32_BIT 64_BIT
features: COMPRESSED
old_n_functions: 0000
crc16: 0000
ctype:
library_name_len: 25
ctypes_crc16: 0000
ARM library little endian
inflate error: -3 data error
Decompressing failed.
We encountered an error while parsing the file. Sorry.
```

**Comment by b1llow** [comment]
It's OK now. But I just test IDA Pro 7.5.

```
rizin on  billow [!?] is 📦 v0.3.0-git
❯ for a in (find ../sig_75/ -name \*.sig); build/binrz/rizin/rizin -qc "zfd $a" 2>&1 | grep 'We encountered an error' ; end

rizin on  billow [!?] is 📦 v0.3.0-git took 22s
❯ for a in (find ../sig_75/ -name \*.sig); build/binrz/rizin/rizin -qc "zfd $a" 2>&1 | grep 'runtime error' ; end
```

**Comment by XVilka** [comment]
@imbillow this one UBSAN runtime error still happens, please take a look too. The rest is good:
```
IDA/ida-7.6-sp1/sig took 1m 2s 
[i] ℤ for a in $(find . -name *.sig); rizin -qc "zfd $a" 2>&1 | grep 'runtime error'                                                                                                                                              13:23:51 
../librz/analysis/flirt.c:383:26: runtime error: left shift of 52428 by 16 places cannot be represented in type 'int'
IDA/ida-7.5-sp2/sig 
[i] ℤ for a in $(find . -name *.sig); rizin -qc "zfd $a" 2>&1 | grep 'runtime error'                                                                                                                                              13:24:59 
../librz/analysis/flirt.c:383:26: runtime error: left shift of 52428 by 16 places cannot be represented in type 'int'
IDA/ida-7.2-181105/sig took 1m 0s 
[i] ℤ for a in $(find . -name *.sig); rizin -qc "zfd $a" 2>&1 | grep 'runtime error'                                                                                                                                              13:23:38 
../librz/analysis/flirt.c:383:26: runtime error: left shift of 52428 by 16 places cannot be represented in type 'int'
ida-7.6-sp1/sig/mips took 8s 
[i] ℤ rizin =                                                                                                                                                                                                                     
[0x00000000]> zfd psx.sig
../librz/analysis/flirt.c:383:26: runtime error: left shift of 52428 by 16 places cannot be represented in type 'int'
....023C....4224....033C....6324000040AC040042242B084300FCFF2014:
 0. 04 DEFC 00A8 0000:o=20:a=12A:psyq:m=+A0^/main (0087: 00)
 1. 04 DEFC 0180 0000:o=20:a=12A:psyq:m=+9C^/main 00A8:__main 0118:__do_global_dtors (0087: 3C)
[0x00000000]> 
```

**Review by XVilka — APPROVED** [review]

**Pull request merged**